include/sparrow/array_data.hpp

```diff
@@ -36,8 +50,251 @@ namespace sparrow
         std::vector<array_data> child_data;
     };
 
-    struct null_type
+    /**
+     * Reference type used in layout classes.
+     * This classes provides a similar behavior
+     * as std::optional, but on references in
+     * an array_data.
+     */
+    template <class L>
+    class reference_proxy
     {
+    public:
+
+        using self_type = reference_proxy<L>;
+        using layout_type = L;
+        using value_type = typename L::inner_value_type;
+        using size_type = typename L::size_type;
+
+        reference_proxy(layout_type& l, size_type index);
+
+        bool has_value() const;
+        explicit operator bool() const;
+
+        value_type& value();
+        const value_type& value() const;
+
+        self_type& operator=(const self_type& rhs);
+        self_type& operator=(self_type&& rhs);
+        self_type& operator=(std::nullopt_t);
+
+        template <std::convertible_to<typename L::inner_value_type> U>
+        self_type& operator=(const std::optional<U>& rhs);
+
+        template <std::convertible_to<typename L::inner_value_type> U>
+        self_type& operator=(std::optional<U>&& rhs);
+
+        template <std::convertible_to<typename L::inner_value_type> U>
+        self_type& operator=(U&& value);
+
+        void reset();
+        void swap(self_type& rhs);
+        
```

Comment:
Consider adding a `// TODO: ...` for adding all the other usual arithmetic operators that might be allowed for `T`?

Reply:
They don't exist for `std::optional`, that is why I haven't added them.

Reply:
Yes, I meant do we think we will need them in the context of arrow db manipulation?

Reply:
For now I don't think so. most of the time, people simply perform the operations through the pair of value iterator / bitmask iterator, and perform operations even if the bitmask if false. If your data is not too sparse (which is the case most of the time), it is faster than testing the flag and branching for each value.